test/helpers/block.ts

```diff
@@ -0,0 +1,378 @@
+import "@moonbeam-network/api-augment/moonbase";
+import type { RuntimeDispatchInfoV1 } from "@polkadot/types/interfaces/payment";
+import { ApiPromise } from "@polkadot/api";
+import { BlockHash, DispatchInfo, RuntimeDispatchInfo } from "@polkadot/types/interfaces";
+import { u32, u128, Option } from "@polkadot/types";
+import {
+  WEIGHT_PER_GAS,
+  mapExtrinsics,
+  BlockRangeOption,
+  calculateFeePortions,
+  EXTRINSIC_BASE_WEIGHT,
+} from "@moonwall/util";
+import { DevModeContext } from "@moonwall/cli";
+import { expect } from "@moonwall/cli";
+import type { Block, AccountId20 } from "@polkadot/types/interfaces/runtime/types";
+import type { TxWithEvent } from "@polkadot/api-derive/types";
+import type { ITuple } from "@polkadot/types-codec/types";
+
+const debug = require("debug")("test:blocks");
+
+// Given a deposit amount, returns the amount burned (80%) and deposited to treasury (20%).
+// This is meant to precisely mimic the logic in the Moonbeam runtimes where the burn amount
+// is calculated and the treasury is treated as the remainder. This precision is important to
+// avoid off-by-one errors.
+
+export interface TxWithEventAndFee extends TxWithEvent {
+  fee: RuntimeDispatchInfo | RuntimeDispatchInfoV1;
+}
+
+export interface BlockDetails {
+  block: Block;
+  txWithEvents: TxWithEventAndFee[];
+}
+
+const getBlockDetails = async (
+  api: ApiPromise,
+  blockHash: BlockHash | string | any
+): Promise<BlockDetails> => {
+  debug(`Querying ${blockHash}`);
+
+  const [{ block }, records] = await Promise.all([
+    api.rpc.chain.getBlock(blockHash),
+    await (await api.at(blockHash)).query.system.events(),
+  ]);
+
+  const fees = await Promise.all(
+    block.extrinsics.map((ext) => api.rpc.payment.queryInfo(ext.toHex(), block.header.parentHash))
+  );
+
+  const txWithEvents = mapExtrinsics(block.extrinsics, records, fees);
+
+  return {
+    block,
+    txWithEvents,
+  } as any as BlockDetails;
+};
+
+// Explore all blocks for the given range and returns block information for each one
+// fromBlockNumber and toBlockNumber included
+export const exploreBlockRange = async (
+  api: ApiPromise,
+  { from, to, concurrency = 1 }: BlockRangeOption,
+  callBack: (blockDetails: BlockDetails) => Promise<void>
+) => {
+  let current = from;
+  while (current <= to) {
+    const concurrentTasks: any[] = [];
+    for (let i = 0; i < concurrency && current <= to; i++) {
+      concurrentTasks.push(
+        api.rpc.chain.getBlockHash(current++).then((hash) => getBlockDetails(api, hash))
+      );
+    }
+    const blocksDetails = await Promise.all(concurrentTasks);
+    for (const blockDetails of blocksDetails) {
+      await callBack(blockDetails);
+    }
+  }
+};
+
+export const verifyBlockFees = async (
```

Comment:
This could get moved to moonbeam-tools or something specific, as it has been copied to few places and is not maintained in all of them